<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • AlphaSimR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">AlphaSimR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/AlphaSimR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/GeneticRecombination.html">Genetic Recombination</a></li>
    <li><a class="dropdown-item" href="../articles/Genome.html">The Genome</a></li>
    <li><a class="dropdown-item" href="../articles/traits.html">Traits in AlphaSimR</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/gaynorr/AlphaSimR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Changelog</h1>
      <small>Source: <a href="https://github.com/gaynorr/AlphaSimR/blob/master/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.1.0" id="alphasimr-210">AlphaSimR 2.1.0<a class="anchor" aria-label="anchor" href="#alphasimr-210"></a></h2><p class="text-muted">CRAN release: 2025-11-08</p>
<p>*changed R6 and methods from Depends to Imports to match current best practices for R packages</p>
<p>*Change order of call to <code>finalizePop</code> in <code>.newPop</code> to allow access to recombination tracking data</p>
<p>*added <code>parentAverage</code> and <code>mendelianSampling</code> functions</p>
<p>*fixed bug in <code>c</code> for RawPop, MapPop, and NamedMapPop</p>
<p>*changed <code>popVar</code> to an R wrapper to automate casting of vectors to matrices</p>
<p>*corrected bibliography entry month from sept to sep</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.0.0" id="alphasimr-200">AlphaSimR 2.0.0<a class="anchor" aria-label="anchor" href="#alphasimr-200"></a></h2><p class="text-muted">CRAN release: 2025-09-01</p>
<p>*added names to <code>SP$recHist</code></p>
<p>*added <code>asCategorical</code> to convert a normal (Gaussian) trait to an ordered categorical (threshold) trait</p>
<p>*improved computational performance of simulations with multiple traits</p>
<p>*added support for data.frames in SimParam genetic map switching functions</p>
<p>*changed finalizePop function call in <code>.newPop</code> to pass simParam as an argument</p>
<p>*updated version numbering to follow tidyverse format with a major version indicating backwards compatibility has been broken</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.6.1" id="alphasimr-161">AlphaSimR 1.6.1<a class="anchor" aria-label="anchor" href="#alphasimr-161"></a></h2><p class="text-muted">CRAN release: 2024-11-01</p>
<p>*fixed bug in <code>mergePops</code> and <code>[</code> (subset) methods - they were failing for populations that had a misc slot with a matrix - we now check if a misc slot element is a matrix and rbind them for <code>mergePops</code> and subset rows for <code>[</code> (assuming the first dimension represents individuals)</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.6.0" id="alphasimr-160">AlphaSimR 1.6.0<a class="anchor" aria-label="anchor" href="#alphasimr-160"></a></h2><p class="text-muted">CRAN release: 2024-08-17</p>
<p>*exported <code>meanEBV</code> and added <code>varEBV</code> to complement <code>meanP</code>/<code>varP</code> and <code>meanG</code>/<code>varG</code></p>
<p>*Changed all parameters of the CATTLE demographic model to exactly match Macleod et al. (2013) - specifically reducing the mutation rate from 2.5e-8 (from human literature) to 1.2e-8 (used in Macleod et al., 2013) and recombination rate from 1e-8 (generic) to 9.26e-9 (used in Macleod et al., 2013). These changes will reduce number of segregating sites to ~240K per chromosome for 100 samples and will run faster.</p>
<p>*changed misc slot in Pop class from a list organised as ind x nodes to to a list organised as nodes x ind (this simplified code and increased speed)</p>
<p>*removed <code>setMisc</code> and <code>getMisc</code> because the new misc slot structure makes it easy to set and get misc components with base R code</p>
<p>*added <code>length</code> method for Pop class that returns number of individuals (like <code>nInd</code>)</p>
<p>*added <code>length</code> method for MultiPop class that returns number of populations</p>
<p>*fixed bug in quadrivalent pairing resulting in distribution of double reductions not respecting the centromere</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.5.3" id="alphasimr-153">AlphaSimR 1.5.3<a class="anchor" aria-label="anchor" href="#alphasimr-153"></a></h2><p class="text-muted">CRAN release: 2023-12-01</p>
<p>*fixed bug in <code>SimParam$restrSegSites</code> with excluding sites at end of chromosome</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.5.2" id="alphasimr-152">AlphaSimR 1.5.2<a class="anchor" aria-label="anchor" href="#alphasimr-152"></a></h2><p class="text-muted">CRAN release: 2023-11-01</p>
<p>*fix SimParam examples for CRAN</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.5.1" id="alphasimr-151">AlphaSimR 1.5.1<a class="anchor" aria-label="anchor" href="#alphasimr-151"></a></h2>
<p>*deleted bad example code for <code>setMisc</code></p>
<p>*changed examples to use a single thread for CRAN testing this change is not shown in the documentation</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.5.0" id="alphasimr-150">AlphaSimR 1.5.0<a class="anchor" aria-label="anchor" href="#alphasimr-150"></a></h2>
<p>*renamed <code>MegaPop</code> to <code>MultiPop</code></p>
<p>*fixed bug in <code>writePlink</code> to correctly export map positions in cM</p>
<p>*fixed bug in <code>writeRecords</code> due to removed reps slot in pops</p>
<p>*added <code>altAddTraitAD</code> for specifying traits with dominance effects using dominance variance and inbreeding depression</p>
<p>*add miscPop slot to class <code>Pop</code></p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.4.2" id="alphasimr-142">AlphaSimR 1.4.2<a class="anchor" aria-label="anchor" href="#alphasimr-142"></a></h2><p class="text-muted">CRAN release: 2023-03-27</p>
<p>*updated MaCS citation to https site</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.4.1" id="alphasimr-141">AlphaSimR 1.4.1<a class="anchor" aria-label="anchor" href="#alphasimr-141"></a></h2>
<p>*Changed citation to use <code>bibentry</code> instead of <code>citEntry</code></p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.4.0" id="alphasimr-140">AlphaSimR 1.4.0<a class="anchor" aria-label="anchor" href="#alphasimr-140"></a></h2>
<p>*fixed a bug in IBD tracking</p>
<p>*add <code>setFounderHap</code> to SimParam for applying custom haplotypes to founders</p>
<p>*added <code>addSnpChipByName</code> to SimParam for defining SNP chips by marker names</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.4" id="alphasimr-134">AlphaSimR 1.3.4<a class="anchor" aria-label="anchor" href="#alphasimr-134"></a></h2><p class="text-muted">CRAN release: 2022-12-09</p>
<p>*changed C++ using <code>sprintf</code> to use <code>snprintf</code></p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.3" id="alphasimr-133">AlphaSimR 1.3.3<a class="anchor" aria-label="anchor" href="#alphasimr-133"></a></h2>
<p>*fixed bug in calculation of genic variance</p>
<p>*fixed <code>importHaplo</code> not passing ploidy to <code>newMapPop</code></p>
<p>*fixed bug with correlated error variances</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.2" id="alphasimr-132">AlphaSimR 1.3.2<a class="anchor" aria-label="anchor" href="#alphasimr-132"></a></h2><p class="text-muted">CRAN release: 2022-11-02</p>
<p>*fixed column name bug with multiple traits in <code>setEBV</code></p>
<p>*fixed CTD caused by <code>runMacs</code> when too many segSites are requested</p>
<p>*fixed missing names in GV when using <code>resetPop</code></p>
<p>*fixed bug in <code>importTrait</code></p>
<p>*<code>popVar</code> now deals with matrices having 1 row</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.1" id="alphasimr-131">AlphaSimR 1.3.1<a class="anchor" aria-label="anchor" href="#alphasimr-131"></a></h2><p class="text-muted">CRAN release: 2022-08-25</p>
<p>*updated link to Gaynor, 2017</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.0" id="alphasimr-130">AlphaSimR 1.3.0<a class="anchor" aria-label="anchor" href="#alphasimr-130"></a></h2>
<p>*added ability to exclude loci by name in <code>SimParam$restrSegSites</code></p>
<p>*<code>pullMarkerGeno</code> and <code>pullMarkerHaplo</code> now work with a MapPop class</p>
<p>*added <code>setMarkerHaplo</code> to manually change genotypes in a Pop or MapPop</p>
<p>*added <code>addSegSite</code> for manually adding segregating sites to a MapPop class</p>
<p>*<code>simParam$setCorE</code> has been deprecated in favor of a corE argument in <code>simParam$setVarE</code></p>
<p>*<code>setPheno</code> now takes corE as an argument</p>
<p>*<code>setPheno</code> now allows the user to set phenotypes for a subset of traits</p>
<p>*add <code>newEmptyPop</code> to create populations with zero individuals</p>
<p>*removed reps slot from populations and heterogeneous residual variance GS models</p>
<p>*added h2, H2, and corE to <code>setPhenoGCA</code> and <code>setPhenoProgTest</code></p>
<p>*the “EUROPEAN” species history was removed from <code>runMacs</code> due to lengthy runtime</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.2.2" id="alphasimr-122">AlphaSimR 1.2.2<a class="anchor" aria-label="anchor" href="#alphasimr-122"></a></h2>
<p>*added <code>getPed</code> to quick extract a population’s pedigree</p>
<p>*added <code>getGenMap</code> to pull a genetic map in data.frame format</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.2.1" id="alphasimr-121">AlphaSimR 1.2.1<a class="anchor" aria-label="anchor" href="#alphasimr-121"></a></h2><p class="text-muted">CRAN release: 2022-07-05</p>
<p>*fixed bugs relating to <code>importData</code> functions</p>
<p>*fixed <code>writePlink</code> errors and no longer requires equal length chromosomes</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.2.0" id="alphasimr-120">AlphaSimR 1.2.0<a class="anchor" aria-label="anchor" href="#alphasimr-120"></a></h2>
<p>*added <code>importGenMap</code> to format genetic maps for AlphaSimR</p>
<p>*added <code>importInbredGeno</code> and <code>importHaplo</code> to make it easier to create a simulation from external data</p>
<p>*added <code>importSnpChip</code>, <code>importTrait</code> to <code>SimParam</code> to make it easier to manually define traits</p>
<p>*added <code>pullMarkerGeno</code> and <code>pullMarkerHaplo</code> to make it easier to extract genotypes and haplotypes of specific loci without defining a trait or SNP chip</p>
<p>*<code>reduceGenome</code>, <code>mergeGenome</code> and <code>doubleGenome</code> should really now work with pedigree and recombination tracking</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.2" id="alphasimr-112">AlphaSimR 1.1.2<a class="anchor" aria-label="anchor" href="#alphasimr-112"></a></h2><p class="text-muted">CRAN release: 2022-02-17</p>
<p>*added missing #ifdef _OPENMP to OCS.cpp</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.1" id="alphasimr-111">AlphaSimR 1.1.1<a class="anchor" aria-label="anchor" href="#alphasimr-111"></a></h2>
<p>*removed use of PI variable in C++ code due to it being compiler specific</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.0" id="alphasimr-110">AlphaSimR 1.1.0<a class="anchor" aria-label="anchor" href="#alphasimr-110"></a></h2>
<p>*added snpChip argument to <code>pullIbdHaplo</code> for backwards compatibility</p>
<p>*exposed internal mixed model solvers</p>
<p>*all selection functions now return a warning when there are not enough individuals</p>
<p>*fixed error in <code>pullIbdHaplo</code> when chr isn’t NULL</p>
<p>*fixed an error with assigning 1 QTL and/or SNP</p>
<p>*changed geno slot from matrix to list to support future RcppArmadillo changes</p>
<p>*<code>doubleGenome</code> and <code>reduceGenome</code> now work with IBD tracking</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.4" id="alphasimr-104">AlphaSimR 1.0.4<a class="anchor" aria-label="anchor" href="#alphasimr-104"></a></h2><p class="text-muted">CRAN release: 2021-09-08</p>
<p>*fixed errors in implementation of Gamma Sprinkling model</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.3" id="alphasimr-103">AlphaSimR 1.0.3<a class="anchor" aria-label="anchor" href="#alphasimr-103"></a></h2><p class="text-muted">CRAN release: 2021-07-29</p>
<p>*fixed formatting error in genetic maps created by runMacs that broke genotype extraction functions</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.2" id="alphasimr-102">AlphaSimR 1.0.2<a class="anchor" aria-label="anchor" href="#alphasimr-102"></a></h2><p class="text-muted">CRAN release: 2021-07-28</p>
<p>*added h2 and H2 to <code>setPhenoGCA</code></p>
<p>*<code>pullGeno</code> and <code>pullHaplo</code> functions now report marker names from the genetic map</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.1" id="alphasimr-101">AlphaSimR 1.0.1<a class="anchor" aria-label="anchor" href="#alphasimr-101"></a></h2><p class="text-muted">CRAN release: 2021-05-14</p>
<p>*removed lazyData field in DESCRIPTION</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.0" id="alphasimr-100">AlphaSimR 1.0.0<a class="anchor" aria-label="anchor" href="#alphasimr-100"></a></h2>
<p>*AlphaSimR manuscript has been published in G3 (citation added)</p>
<p>*changed to a Gamma Sprinkling model for crossovers, default is still a Gamma model</p>
<p>*change default interference parameter (v) to 2.6 to be consistent with the Kosambi mapping function (was 1, consistent with the Haldane mapping function)</p>
<p>*new internal id (iid) that allows user to freely change id slot in populations</p>
<p>*<code>runMacs2</code> now adjusts Ne for autopolyploids</p>
<p>*parent populations are now passed to <code>finalizePop</code></p>
<p>*check added that throws an error when use of discontinued “gender” argument is detected</p>
<p>*added experimental <code>MegaPop-class</code></p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.13.0" id="alphasimr-0130">AlphaSimR 0.13.0<a class="anchor" aria-label="anchor" href="#alphasimr-0130"></a></h2><p class="text-muted">CRAN release: 2020-10-11</p>
<p>*references to gender have been changed to the more appropriate terms sex or sexes</p>
<p>*added misc slot to populations</p>
<p>*added <code>finalizePop</code> to <code>SimParam</code></p>
<p>*added physical positions to <code>getSnpMap</code> and <code>getQtlMap</code></p>
<p>*you can now use h2 and H2 to specify error variance in <code>setPheno</code></p>
<p>*<code>SimParam$setVarE</code> now accepts a matrix for varE</p>
<p>*fixed a bug in <code>editGenome</code> when making multiple edits</p>
<p>*adding merging of centromere vector in <code>cChr</code></p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.12.2" id="alphasimr-0122">AlphaSimR 0.12.2<a class="anchor" aria-label="anchor" href="#alphasimr-0122"></a></h2><p class="text-muted">CRAN release: 2020-05-19</p>
<p>*GxE traits now default to random sampling of p-values</p>
<p>*fixed a bug in <code>restrSegSites</code></p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.12.1" id="alphasimr-0121">AlphaSimR 0.12.1<a class="anchor" aria-label="anchor" href="#alphasimr-0121"></a></h2><p class="text-muted">CRAN release: 2020-03-30</p>
<p>*fixed a bug in selection of segSites</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.12.0" id="alphasimr-0120">AlphaSimR 0.12.0<a class="anchor" aria-label="anchor" href="#alphasimr-0120"></a></h2>
<p>*changed output of <code>genParam</code> to match Bulmer, 1976</p>
<p>*nProgeny added to <code>makeCross</code> and <code>makeCross2</code></p>
<p>*all <code>SimParam</code> documentation is now in <code><a href="../reference/SimParam.html">?SimParam</a></code></p>
<p>*non-overlapping QTL and SNP is now the default</p>
<p>*new interface for <code>restrSegSites</code> in <code>SimParam</code></p>
<p>*fixed subset by id for populations</p>
<p>*fixed major bug in <code>newMapPop</code></p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.11.1" id="alphasimr-0111">AlphaSimR 0.11.1<a class="anchor" aria-label="anchor" href="#alphasimr-0111"></a></h2><p class="text-muted">CRAN release: 2020-01-13</p>
<p>*switched to a circular design for the balance option in <code>randCross</code> and <code>randCross2</code></p>
<p>*added <code>reduceGenome</code> and <code>doubleGenome</code> for changing ploidy levels</p>
<p>*added minSnpFreq to SimParam_addSnpChip for any reference population</p>
<p>*the <code>c</code> function now merges individuals for MapPop objects (was chromosomes before)</p>
<p>*the <code>cChr</code> function new merges chromosomes for MapPop objects</p>
<p>*fixed broken SimParam_addStructuredSnpChip</p>
<p>*removed broken <code>pullMultipleSnpGeno</code> and <code>pullMultipleSnpHaplo</code></p>
<p>*fixed broken <code>writePlink</code></p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.11.0" id="alphasimr-0110">AlphaSimR 0.11.0<a class="anchor" aria-label="anchor" href="#alphasimr-0110"></a></h2><p class="text-muted">CRAN release: 2019-10-11</p>
<p>*rework of <code>setEBV</code> (breaks some scripts)</p>
<p>*genotype data now stored as bits (was bytes)</p>
<p>*implemented a gamma model for crossover interference</p>
<p>*added the mutate function to model random mutations</p>
<p>*added a vignette explaining the biological model for traits</p>
<p>*GS models now handle polyploids</p>
<p>*heterogenous error variance is now optional in GS models (default is homogeneous error)</p>
<p>*improved gene drop functionality of pedigreeCross</p>
<p>*added keepParents option to makeDH and self (indirectly extends <code>selectFam</code> and <code>selectWithinFam</code>)</p>
<p>*added RRBLUP_SCA2</p>
<p>*set methods for the “show” function when applied to populations</p>
<p>*fixed a bug returning the first individual when selecting 0</p>
<p>*fixed error in recombination track when using <code>makeDH</code></p>
<p>*fixed error causing epistatic effects to mask GxE effects</p>
<p>*fixed an error with <code>pullSegSiteGeno</code> and <code>pullSegSiteHaplo</code> with variable number of sites per chromosome</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.10.0" id="alphasimr-0100">AlphaSimR 0.10.0<a class="anchor" aria-label="anchor" href="#alphasimr-0100"></a></h2><p class="text-muted">CRAN release: 2019-03-15</p>
<p>*added traits with epistasis</p>
<p>*Max number of threads automatically detected</p>
<p>*added RRBLUP_D2</p>
<p>*added version tracking to <code>SimParam</code></p>
<p>*removed <code>trackHaploPop</code> (super-ceded by <code>pullIbdHaplo</code>)</p>
<p>*added <code>fastRRBLUP</code></p>
<p>*fixed faulty double crossover logic</p>
<p>*fixed broken <code>writePlink</code></p>
<p>*fixed broken <code>pullIbdHaplo</code></p>
<p>*<code>mergePops</code> no longer assumes diploidy</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.0" id="alphasimr-090">AlphaSimR 0.9.0<a class="anchor" aria-label="anchor" href="#alphasimr-090"></a></h2><p class="text-muted">CRAN release: 2019-02-07</p>
<p>*added support for autopolyploids</p>
<p>*added <code>RRBLUP_GCA2</code></p>
<p>*<code>randCross2</code> can now “balance” crossing when not using gender</p>
<p>*fixed recombination tracking bug in <code>createDH2</code></p>
<p>*removed bug in <code>setEBV</code> with append=TRUE</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.8.2" id="alphasimr-082">AlphaSimR 0.8.2<a class="anchor" aria-label="anchor" href="#alphasimr-082"></a></h2><p class="text-muted">CRAN release: 2018-10-15</p>
<p>*fixed ambiguous overloading in optimize.cpp</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.8.1" id="alphasimr-081">AlphaSimR 0.8.1<a class="anchor" aria-label="anchor" href="#alphasimr-081"></a></h2><p class="text-muted">CRAN release: 2018-10-12</p>
<p>*<code>setPheno</code> (not <code>setPhenoGCA</code>) passes the number of reps to populations</p>
<p>*fixed bug in <code>editGenomeTopQtl</code></p>
<p>*fixed bug in <code>RRBLUP_D</code></p>
<p>*fixed bug in <code>resetPop</code></p>
<p>*fixed bug in SimParam_rescaleTraits</p>
<p>*removed unimplemented SimParam_restrSnpSites and SimParam_restrQtlSites</p>
<p>*add error message for no traits in <code>calcGCA</code></p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.8.0" id="alphasimr-080">AlphaSimR 0.8.0<a class="anchor" aria-label="anchor" href="#alphasimr-080"></a></h2>
<p>*added GxE traits with zero environmental variance</p>
<p>*faster trait scaling</p>
<p>*faster calculation of genetic values</p>
<p>*dsyevr now called via arma_fortran</p>
<p>*added OpenMP support</p>
<p>*parallelized <code>cross2</code></p>
<p>*parallelized <code>runMacs</code></p>
<p>*parallelized calculation of genetic values</p>
<p>*variance calculations now account for inbreeding</p>
<p>*fixes for male selection in <code>selectOP</code></p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.7.1" id="alphasimr-071">AlphaSimR 0.7.1<a class="anchor" aria-label="anchor" href="#alphasimr-071"></a></h2>
<p>*add fixEff to <code>setPhenoGCA</code></p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.7.0" id="alphasimr-070">AlphaSimR 0.7.0<a class="anchor" aria-label="anchor" href="#alphasimr-070"></a></h2>
<p>*added default <code>runMacs</code> option to return all segSites</p>
<p>*added ability to specify separate male and female genetic maps</p>
<p>*<code>pullGeno</code> and <code>pullHaplo</code> functions can now specify chromosomes</p>
<p>*added <code>RRBLUP2</code> for special GS cases</p>
<p>*improved speed by replacing Rcpp random number generators</p>
<p>*changed available MaCS species</p>
<p>*GS functions now use populations directly</p>
<p>*added <code>pullIbdHaplo</code></p>
<p>*added <code>writePlink</code></p>
<p>*fixed population sub-setting checks to prevent invalid selections</p>
<p>*fixed slow <code>calcGCA</code></p>
<p>*fixed error in <code>addTraitAG</code> preventing multiple traits</p>
<p>*fixed bug with <code>mergePops</code> when merging ebv</p>
<p>*fixed bug in <code>setVarE</code> when using H2 and multiple traits</p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.1" id="alphasimr-061">AlphaSimR 0.6.1<a class="anchor" aria-label="anchor" href="#alphasimr-061"></a></h2>
<p>*<code>selectFam</code> now handles half-sib families</p>
<p>*<code>selectWithinFam</code> now handles half-sib families</p>
<p>*Removed restriction on varE=NULL in <code>setPhenoGCA</code></p>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.0" id="alphasimr-060">AlphaSimR 0.6.0<a class="anchor" aria-label="anchor" href="#alphasimr-060"></a></h2>
<p>*Added NEWS file</p>
<p>*Added <code>selectOP</code> to model selection in open pollinating plants</p>
<p>*Added <code>runMacs2</code> as a wrapper for <code>runMacs</code></p>
<p>*Fixed error when using H2 in SimParam_setVarE</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Gaynor.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

