<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ AlphaSimR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">AlphaSimR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.6.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/AlphaSimR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/DataImport.html">Data Import</a></li>
    <li><a class="dropdown-item" href="../articles/Genome.html">Genome</a></li>
    <li><a class="dropdown-item" href="../articles/GenomicSelection.html">Genomic Selection</a></li>
    <li><a class="dropdown-item" href="../articles/intro.html">Introduction to AlphaSimR</a></li>
    <li><a class="dropdown-item" href="../articles/Meiosis.html">Meiosis</a></li>
    <li><a class="dropdown-item" href="../articles/Misc.html">Misc</a></li>
    <li><a class="dropdown-item" href="../articles/QuantGen.html">Quantitative Genetics</a></li>
    <li><a class="dropdown-item" href="../articles/traits.html">Traits in AlphaSimR</a></li>
    <li><a class="dropdown-item" href="../articles/Traits.html">Traits</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/gaynorr/AlphaSimR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="HybridPop-class.html">`[`(<i>&lt;HybridPop&gt;</i>)</a></code> <code><a href="HybridPop-class.html">c(<i>&lt;HybridPop&gt;</i>)</a></code> <code><a href="HybridPop-class.html">isHybridPop()</a></code>

        </dt>
        <dd>Hybrid population</dd>
      </dl><dl><dt>

          <code><a href="LociMap-class.html">LociMap-class</a></code>

        </dt>
        <dd>Loci metadata</dd>
      </dl><dl><dt>

          <code><a href="MapPop-class.html">`[`(<i>&lt;MapPop&gt;</i>)</a></code> <code><a href="MapPop-class.html">c(<i>&lt;MapPop&gt;</i>)</a></code> <code><a href="MapPop-class.html">isMapPop()</a></code>

        </dt>
        <dd>Raw population with genetic map</dd>
      </dl><dl><dt>

          <code><a href="MultiPop-class.html">`[`(<i>&lt;MultiPop&gt;</i>)</a></code> <code><a href="MultiPop-class.html">`[[`(<i>&lt;MultiPop&gt;</i>)</a></code> <code><a href="MultiPop-class.html">c(<i>&lt;MultiPop&gt;</i>)</a></code> <code><a href="MultiPop-class.html">length(<i>&lt;MultiPop&gt;</i>)</a></code> <code><a href="MultiPop-class.html">isMultiPop()</a></code>

        </dt>
        <dd>Multi-Population</dd>
      </dl><dl><dt>

          <code><a href="NamedMapPop-class.html">`[`(<i>&lt;NamedMapPop&gt;</i>)</a></code> <code><a href="NamedMapPop-class.html">c(<i>&lt;NamedMapPop&gt;</i>)</a></code> <code><a href="NamedMapPop-class.html">isNamedMapPop()</a></code>

        </dt>
        <dd>Raw population with genetic map and id</dd>
      </dl><dl><dt>

          <code><a href="Pop-class.html">`[`(<i>&lt;Pop&gt;</i>)</a></code> <code><a href="Pop-class.html">c(<i>&lt;Pop&gt;</i>)</a></code> <code><a href="Pop-class.html">show(<i>&lt;Pop&gt;</i>)</a></code> <code><a href="Pop-class.html">length(<i>&lt;Pop&gt;</i>)</a></code>

        </dt>
        <dd>Population</dd>
      </dl><dl><dt>

          <code><a href="RRBLUP.html">RRBLUP()</a></code>

        </dt>
        <dd>RR-BLUP Model</dd>
      </dl><dl><dt>

          <code><a href="RRBLUP2.html">RRBLUP2()</a></code>

        </dt>
        <dd>RR-BLUP Model 2</dd>
      </dl><dl><dt>

          <code><a href="RRBLUPMemUse.html">RRBLUPMemUse()</a></code>

        </dt>
        <dd>RRBLUP Memory Usage</dd>
      </dl><dl><dt>

          <code><a href="RRBLUP_D.html">RRBLUP_D()</a></code>

        </dt>
        <dd>RR-BLUP Model with Dominance</dd>
      </dl><dl><dt>

          <code><a href="RRBLUP_D2.html">RRBLUP_D2()</a></code>

        </dt>
        <dd>RR-BLUP with Dominance Model 2</dd>
      </dl><dl><dt>

          <code><a href="RRBLUP_GCA.html">RRBLUP_GCA()</a></code>

        </dt>
        <dd>RR-BLUP GCA Model</dd>
      </dl><dl><dt>

          <code><a href="RRBLUP_GCA2.html">RRBLUP_GCA2()</a></code>

        </dt>
        <dd>RR-BLUP GCA Model 2</dd>
      </dl><dl><dt>

          <code><a href="RRBLUP_SCA.html">RRBLUP_SCA()</a></code>

        </dt>
        <dd>RR-BLUP SCA Model</dd>
      </dl><dl><dt>

          <code><a href="RRBLUP_SCA2.html">RRBLUP_SCA2()</a></code>

        </dt>
        <dd>RR-BLUP SCA Model 2</dd>
      </dl><dl><dt>

          <code><a href="RRsol-class.html">RRsol-class</a></code>

        </dt>
        <dd>RR-BLUP Solution</dd>
      </dl><dl><dt>

          <code><a href="RawPop-class.html">`[`(<i>&lt;RawPop&gt;</i>)</a></code> <code><a href="RawPop-class.html">c(<i>&lt;RawPop&gt;</i>)</a></code> <code><a href="RawPop-class.html">show(<i>&lt;RawPop&gt;</i>)</a></code> <code><a href="RawPop-class.html">isRawPop()</a></code>

        </dt>
        <dd>Raw Population</dd>
      </dl><dl><dt>

          <code><a href="SimParam.html">SimParam</a></code>

        </dt>
        <dd>Simulation parameters</dd>
      </dl><dl><dt>

          <code><a href="TraitA-class.html">TraitA-class</a></code>

        </dt>
        <dd>Additive trait</dd>
      </dl><dl><dt>

          <code><a href="TraitA2-class.html">TraitA2-class</a></code>

        </dt>
        <dd>Sex specific additive trait</dd>
      </dl><dl><dt>

          <code><a href="TraitA2D-class.html">TraitA2D-class</a></code>

        </dt>
        <dd>Sex specific additive and dominance trait</dd>
      </dl><dl><dt>

          <code><a href="TraitAD-class.html">TraitAD-class</a></code>

        </dt>
        <dd>Additive and dominance trait</dd>
      </dl><dl><dt>

          <code><a href="TraitADE-class.html">TraitADE-class</a></code>

        </dt>
        <dd>Additive, dominance, and epistatic trait</dd>
      </dl><dl><dt>

          <code><a href="TraitADEG-class.html">TraitADEG-class</a></code>

        </dt>
        <dd>Additive, dominance, epistasis, and GxE trait</dd>
      </dl><dl><dt>

          <code><a href="TraitADG-class.html">TraitADG-class</a></code>

        </dt>
        <dd>Additive, dominance and GxE trait</dd>
      </dl><dl><dt>

          <code><a href="TraitAE-class.html">TraitAE-class</a></code>

        </dt>
        <dd>Additive and epistatic trait</dd>
      </dl><dl><dt>

          <code><a href="TraitAEG-class.html">TraitAEG-class</a></code>

        </dt>
        <dd>Additive, epistasis and GxE trait</dd>
      </dl><dl><dt>

          <code><a href="TraitAG-class.html">TraitAG-class</a></code>

        </dt>
        <dd>Additive and GxE trait</dd>
      </dl><dl><dt>

          <code><a href="aa.html">aa()</a></code>

        </dt>
        <dd>Additive-by-additive epistatic deviations</dd>
      </dl><dl><dt>

          <code><a href="addSegSite.html">addSegSite()</a></code>

        </dt>
        <dd>Add segregating site to MapPop</dd>
      </dl><dl><dt>

          <code><a href="attrition.html">attrition()</a></code>

        </dt>
        <dd>Lose individuals at random</dd>
      </dl><dl><dt>

          <code><a href="bv.html">bv()</a></code>

        </dt>
        <dd>Breeding value</dd>
      </dl><dl><dt>

          <code><a href="cChr.html">cChr()</a></code>

        </dt>
        <dd>Combine MapPop chromosomes</dd>
      </dl><dl><dt>

          <code><a href="calcGCA.html">calcGCA()</a></code>

        </dt>
        <dd>Calculate GCA</dd>
      </dl><dl><dt>

          <code><a href="dd.html">dd()</a></code>

        </dt>
        <dd>Dominance deviations</dd>
      </dl><dl><dt>

          <code><a href="dot-newPop.html">.newPop()</a></code>

        </dt>
        <dd>Create new population (internal)</dd>
      </dl><dl><dt>

          <code><a href="doubleGenome.html">doubleGenome()</a></code>

        </dt>
        <dd>Double the ploidy of individuals</dd>
      </dl><dl><dt>

          <code><a href="ebv.html">ebv()</a></code>

        </dt>
        <dd>Estimated breeding value</dd>
      </dl><dl><dt>

          <code><a href="editGenome.html">editGenome()</a></code>

        </dt>
        <dd>Edit genome</dd>
      </dl><dl><dt>

          <code><a href="editGenomeTopQtl.html">editGenomeTopQtl()</a></code>

        </dt>
        <dd>Edit genome - the top QTL</dd>
      </dl><dl><dt>

          <code><a href="fastRRBLUP.html">fastRRBLUP()</a></code>

        </dt>
        <dd>Fast RR-BLUP</dd>
      </dl><dl><dt>

          <code><a href="genParam.html">genParam()</a></code>

        </dt>
        <dd>Sumarize genetic parameters</dd>
      </dl><dl><dt>

          <code><a href="genicVarA.html">genicVarA()</a></code>

        </dt>
        <dd>Additive genic variance</dd>
      </dl><dl><dt>

          <code><a href="genicVarAA.html">genicVarAA()</a></code>

        </dt>
        <dd>Additive-by-additive genic variance</dd>
      </dl><dl><dt>

          <code><a href="genicVarD.html">genicVarD()</a></code>

        </dt>
        <dd>Dominance genic variance</dd>
      </dl><dl><dt>

          <code><a href="genicVarG.html">genicVarG()</a></code>

        </dt>
        <dd>Total genic variance</dd>
      </dl><dl><dt>

          <code><a href="getGenMap.html">getGenMap()</a></code>

        </dt>
        <dd>Get genetic map</dd>
      </dl><dl><dt>

          <code><a href="getNumThreads.html">getNumThreads()</a></code>

        </dt>
        <dd>Number of available threads</dd>
      </dl><dl><dt>

          <code><a href="getPed.html">getPed()</a></code>

        </dt>
        <dd>Get pedigree</dd>
      </dl><dl><dt>

          <code><a href="getQtlMap.html">getQtlMap()</a></code>

        </dt>
        <dd>Get QTL genetic map</dd>
      </dl><dl><dt>

          <code><a href="getSnpMap.html">getSnpMap()</a></code>

        </dt>
        <dd>Get SNP genetic map</dd>
      </dl><dl><dt>

          <code><a href="gv.html">gv()</a></code>

        </dt>
        <dd>Genetic value</dd>
      </dl><dl><dt>

          <code><a href="hybridCross.html">hybridCross()</a></code>

        </dt>
        <dd>Hybrid crossing</dd>
      </dl><dl><dt>

          <code><a href="importGenMap.html">importGenMap()</a></code>

        </dt>
        <dd>Import genetic map</dd>
      </dl><dl><dt>

          <code><a href="importHaplo.html">importHaplo()</a></code>

        </dt>
        <dd>Import haplotypes</dd>
      </dl><dl><dt>

          <code><a href="importInbredGeno.html">importInbredGeno()</a></code>

        </dt>
        <dd>Import inbred, diploid genotypes</dd>
      </dl><dl><dt>

          <code><a href="isFemale.html">isFemale()</a></code> <code><a href="isFemale.html">isMale()</a></code>

        </dt>
        <dd>Test if individuals of a population are female or male</dd>
      </dl><dl><dt>

          <code><a href="isPop.html">isPop()</a></code>

        </dt>
        <dd>Test if object is of a Population class</dd>
      </dl><dl><dt>

          <code><a href="makeCross.html">makeCross()</a></code>

        </dt>
        <dd>Make designed crosses</dd>
      </dl><dl><dt>

          <code><a href="makeCross2.html">makeCross2()</a></code>

        </dt>
        <dd>Make designed crosses</dd>
      </dl><dl><dt>

          <code><a href="makeDH.html">makeDH()</a></code>

        </dt>
        <dd>Generates DH lines</dd>
      </dl><dl><dt>

          <code><a href="meanEBV.html">meanEBV()</a></code>

        </dt>
        <dd>Mean estimated breeding values</dd>
      </dl><dl><dt>

          <code><a href="meanG.html">meanG()</a></code>

        </dt>
        <dd>Mean genetic values</dd>
      </dl><dl><dt>

          <code><a href="meanP.html">meanP()</a></code>

        </dt>
        <dd>Mean phenotypic values</dd>
      </dl><dl><dt>

          <code><a href="mergeGenome.html">mergeGenome()</a></code>

        </dt>
        <dd>Combine genomes of individuals</dd>
      </dl><dl><dt>

          <code><a href="mergePops.html">mergePops()</a></code>

        </dt>
        <dd>Merge list of populations</dd>
      </dl><dl><dt>

          <code><a href="mutate.html">mutate()</a></code>

        </dt>
        <dd>Add Random Mutations</dd>
      </dl><dl><dt>

          <code><a href="nInd.html">nInd()</a></code>

        </dt>
        <dd>Number of individuals</dd>
      </dl><dl><dt>

          <code><a href="newEmptyPop.html">newEmptyPop()</a></code>

        </dt>
        <dd>Creates an empty population</dd>
      </dl><dl><dt>

          <code><a href="newMapPop.html">newMapPop()</a></code>

        </dt>
        <dd>New MapPop</dd>
      </dl><dl><dt>

          <code><a href="newMultiPop.html">newMultiPop()</a></code>

        </dt>
        <dd>Create new Multi Population</dd>
      </dl><dl><dt>

          <code><a href="newPop.html">newPop()</a></code>

        </dt>
        <dd>Create new population</dd>
      </dl><dl><dt>

          <code><a href="pedigreeCross.html">pedigreeCross()</a></code>

        </dt>
        <dd>Pedigree cross</dd>
      </dl><dl><dt>

          <code><a href="pheno.html">pheno()</a></code>

        </dt>
        <dd>Phenotype</dd>
      </dl><dl><dt>

          <code><a href="popVar.html">popVar()</a></code>

        </dt>
        <dd>Population variance</dd>
      </dl><dl><dt>

          <code><a href="pullIbdHaplo.html">pullIbdHaplo()</a></code>

        </dt>
        <dd>Pull IBD haplotypes</dd>
      </dl><dl><dt>

          <code><a href="pullMarkerGeno.html">pullMarkerGeno()</a></code>

        </dt>
        <dd>Pull marker genotypes</dd>
      </dl><dl><dt>

          <code><a href="pullMarkerHaplo.html">pullMarkerHaplo()</a></code>

        </dt>
        <dd>Pull marker haplotypes</dd>
      </dl><dl><dt>

          <code><a href="pullQtlGeno.html">pullQtlGeno()</a></code>

        </dt>
        <dd>Pull QTL genotypes</dd>
      </dl><dl><dt>

          <code><a href="pullQtlHaplo.html">pullQtlHaplo()</a></code>

        </dt>
        <dd>Pull QTL haplotypes</dd>
      </dl><dl><dt>

          <code><a href="pullSegSiteGeno.html">pullSegSiteGeno()</a></code>

        </dt>
        <dd>Pull segregating site genotypes</dd>
      </dl><dl><dt>

          <code><a href="pullSegSiteHaplo.html">pullSegSiteHaplo()</a></code>

        </dt>
        <dd>Pull seg site haplotypes</dd>
      </dl><dl><dt>

          <code><a href="pullSnpGeno.html">pullSnpGeno()</a></code>

        </dt>
        <dd>Pull SNP genotypes</dd>
      </dl><dl><dt>

          <code><a href="pullSnpHaplo.html">pullSnpHaplo()</a></code>

        </dt>
        <dd>Pull SNP haplotypes</dd>
      </dl><dl><dt>

          <code><a href="quickHaplo.html">quickHaplo()</a></code>

        </dt>
        <dd>Quick founder haplotype simulation</dd>
      </dl><dl><dt>

          <code><a href="randCross.html">randCross()</a></code>

        </dt>
        <dd>Make random crosses</dd>
      </dl><dl><dt>

          <code><a href="randCross2.html">randCross2()</a></code>

        </dt>
        <dd>Make random crosses</dd>
      </dl><dl><dt>

          <code><a href="reduceGenome.html">reduceGenome()</a></code>

        </dt>
        <dd>Create individuals with reduced ploidy</dd>
      </dl><dl><dt>

          <code><a href="resetPop.html">resetPop()</a></code>

        </dt>
        <dd>Reset population</dd>
      </dl><dl><dt>

          <code><a href="runMacs.html">runMacs()</a></code>

        </dt>
        <dd>Create founder haplotypes using MaCS</dd>
      </dl><dl><dt>

          <code><a href="runMacs2.html">runMacs2()</a></code>

        </dt>
        <dd>Alternative wrapper for MaCS</dd>
      </dl><dl><dt>

          <code><a href="sampleHaplo.html">sampleHaplo()</a></code>

        </dt>
        <dd>Sample haplotypes from a MapPop</dd>
      </dl><dl><dt>

          <code><a href="selIndex.html">selIndex()</a></code>

        </dt>
        <dd>Selection index</dd>
      </dl><dl><dt>

          <code><a href="selInt.html">selInt()</a></code>

        </dt>
        <dd>Selection intensity</dd>
      </dl><dl><dt>

          <code><a href="selectCross.html">selectCross()</a></code>

        </dt>
        <dd>Select and randomly cross</dd>
      </dl><dl><dt>

          <code><a href="selectFam.html">selectFam()</a></code>

        </dt>
        <dd>Select families</dd>
      </dl><dl><dt>

          <code><a href="selectInd.html">selectInd()</a></code>

        </dt>
        <dd>Select individuals</dd>
      </dl><dl><dt>

          <code><a href="selectOP.html">selectOP()</a></code>

        </dt>
        <dd>Select open pollinating plants</dd>
      </dl><dl><dt>

          <code><a href="selectWithinFam.html">selectWithinFam()</a></code>

        </dt>
        <dd>Select individuals within families</dd>
      </dl><dl><dt>

          <code><a href="self.html">self()</a></code>

        </dt>
        <dd>Self individuals</dd>
      </dl><dl><dt>

          <code><a href="setEBV.html">setEBV()</a></code>

        </dt>
        <dd>Set estimated breeding values (EBV)</dd>
      </dl><dl><dt>

          <code><a href="setMarkerHaplo.html">setMarkerHaplo()</a></code>

        </dt>
        <dd>Set marker haplotypes</dd>
      </dl><dl><dt>

          <code><a href="setPheno.html">setPheno()</a></code>

        </dt>
        <dd>Set phenotypes</dd>
      </dl><dl><dt>

          <code><a href="setPhenoGCA.html">setPhenoGCA()</a></code>

        </dt>
        <dd>Set GCA as phenotype</dd>
      </dl><dl><dt>

          <code><a href="setPhenoProgTest.html">setPhenoProgTest()</a></code>

        </dt>
        <dd>Set progeny test as phenotype</dd>
      </dl><dl><dt>

          <code><a href="smithHazel.html">smithHazel()</a></code>

        </dt>
        <dd>Calculate Smith-Hazel weights</dd>
      </dl><dl><dt>

          <code><a href="solveMKM.html">solveMKM()</a></code>

        </dt>
        <dd>Solve Multikernel Model</dd>
      </dl><dl><dt>

          <code><a href="solveMVM.html">solveMVM()</a></code>

        </dt>
        <dd>Solve Multivariate Model</dd>
      </dl><dl><dt>

          <code><a href="solveRRBLUP.html">solveRRBLUP()</a></code>

        </dt>
        <dd>Solve RR-BLUP</dd>
      </dl><dl><dt>

          <code><a href="solveRRBLUPMK.html">solveRRBLUPMK()</a></code>

        </dt>
        <dd>Solve Multikernel RR-BLUP</dd>
      </dl><dl><dt>

          <code><a href="solveRRBLUPMV.html">solveRRBLUPMV()</a></code>

        </dt>
        <dd>Solve Multivariate RR-BLUP</dd>
      </dl><dl><dt>

          <code><a href="solveRRBLUP_EM.html">solveRRBLUP_EM()</a></code>

        </dt>
        <dd>Solve RR-BLUP with EM</dd>
      </dl><dl><dt>

          <code><a href="solveRRBLUP_EM2.html">solveRRBLUP_EM2()</a></code>

        </dt>
        <dd>Solve RR-BLUP with EM and 2 random effects</dd>
      </dl><dl><dt>

          <code><a href="solveRRBLUP_EM3.html">solveRRBLUP_EM3()</a></code>

        </dt>
        <dd>Solve RR-BLUP with EM and 3 random effects</dd>
      </dl><dl><dt>

          <code><a href="solveUVM.html">solveUVM()</a></code>

        </dt>
        <dd>Solve Univariate Model</dd>
      </dl><dl><dt>

          <code><a href="transMat.html">transMat()</a></code>

        </dt>
        <dd>Linear transformation matrix</dd>
      </dl><dl><dt>

          <code><a href="usefulness.html">usefulness()</a></code>

        </dt>
        <dd>Usefulness criterion</dd>
      </dl><dl><dt>

          <code><a href="varA.html">varA()</a></code>

        </dt>
        <dd>Additive variance</dd>
      </dl><dl><dt>

          <code><a href="varAA.html">varAA()</a></code>

        </dt>
        <dd>Additive-by-additive epistatic variance</dd>
      </dl><dl><dt>

          <code><a href="varD.html">varD()</a></code>

        </dt>
        <dd>Dominance variance</dd>
      </dl><dl><dt>

          <code><a href="varEBV.html">varEBV()</a></code>

        </dt>
        <dd>Variance of estimated breeding values</dd>
      </dl><dl><dt>

          <code><a href="varG.html">varG()</a></code>

        </dt>
        <dd>Total genetic variance</dd>
      </dl><dl><dt>

          <code><a href="varP.html">varP()</a></code>

        </dt>
        <dd>Phenotypic variance</dd>
      </dl><dl><dt>

          <code><a href="writePlink.html">writePlink()</a></code>

        </dt>
        <dd>Writes a Pop-class as PLINK files</dd>
      </dl><dl><dt>

          <code><a href="writeRecords.html">writeRecords()</a></code>

        </dt>
        <dd>Write data records</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Gaynor.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

