# Package index

## All functions

- [`` `[`( ``*`<HybridPop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/HybridPop-class.md)
  [`c(`*`<HybridPop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/HybridPop-class.md)
  [`isHybridPop()`](https://gaynorr.github.io/AlphaSimR/reference/HybridPop-class.md)
  : Hybrid population
- [`LociMap-class`](https://gaynorr.github.io/AlphaSimR/reference/LociMap-class.md)
  : Loci metadata
- [`` `[`( ``*`<MapPop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/MapPop-class.md)
  [`c(`*`<MapPop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/MapPop-class.md)
  [`isMapPop()`](https://gaynorr.github.io/AlphaSimR/reference/MapPop-class.md)
  : Raw population with genetic map
- [`` `[`( ``*`<MultiPop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/MultiPop-class.md)
  [`` `[[`( ``*`<MultiPop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/MultiPop-class.md)
  [`c(`*`<MultiPop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/MultiPop-class.md)
  [`length(`*`<MultiPop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/MultiPop-class.md)
  [`isMultiPop()`](https://gaynorr.github.io/AlphaSimR/reference/MultiPop-class.md)
  : Multi-Population
- [`` `[`( ``*`<NamedMapPop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/NamedMapPop-class.md)
  [`c(`*`<NamedMapPop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/NamedMapPop-class.md)
  [`isNamedMapPop()`](https://gaynorr.github.io/AlphaSimR/reference/NamedMapPop-class.md)
  : Raw population with genetic map and id
- [`` `[`( ``*`<Pop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/Pop-class.md)
  [`c(`*`<Pop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/Pop-class.md)
  [`show(`*`<Pop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/Pop-class.md)
  [`length(`*`<Pop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/Pop-class.md)
  : Population
- [`RRBLUP()`](https://gaynorr.github.io/AlphaSimR/reference/RRBLUP.md)
  : RR-BLUP Model
- [`RRBLUP2()`](https://gaynorr.github.io/AlphaSimR/reference/RRBLUP2.md)
  : RR-BLUP Model 2
- [`RRBLUPMemUse()`](https://gaynorr.github.io/AlphaSimR/reference/RRBLUPMemUse.md)
  : RRBLUP Memory Usage
- [`RRBLUP_D()`](https://gaynorr.github.io/AlphaSimR/reference/RRBLUP_D.md)
  : RR-BLUP Model with Dominance
- [`RRBLUP_D2()`](https://gaynorr.github.io/AlphaSimR/reference/RRBLUP_D2.md)
  : RR-BLUP with Dominance Model 2
- [`RRBLUP_GCA()`](https://gaynorr.github.io/AlphaSimR/reference/RRBLUP_GCA.md)
  : RR-BLUP GCA Model
- [`RRBLUP_GCA2()`](https://gaynorr.github.io/AlphaSimR/reference/RRBLUP_GCA2.md)
  : RR-BLUP GCA Model 2
- [`RRBLUP_SCA()`](https://gaynorr.github.io/AlphaSimR/reference/RRBLUP_SCA.md)
  : RR-BLUP SCA Model
- [`RRBLUP_SCA2()`](https://gaynorr.github.io/AlphaSimR/reference/RRBLUP_SCA2.md)
  : RR-BLUP SCA Model 2
- [`RRsol-class`](https://gaynorr.github.io/AlphaSimR/reference/RRsol-class.md)
  : RR-BLUP Solution
- [`` `[`( ``*`<RawPop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/RawPop-class.md)
  [`c(`*`<RawPop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/RawPop-class.md)
  [`show(`*`<RawPop>`*`)`](https://gaynorr.github.io/AlphaSimR/reference/RawPop-class.md)
  [`isRawPop()`](https://gaynorr.github.io/AlphaSimR/reference/RawPop-class.md)
  : Raw Population
- [`SimParam`](https://gaynorr.github.io/AlphaSimR/reference/SimParam.md)
  : Simulation parameters
- [`TraitA-class`](https://gaynorr.github.io/AlphaSimR/reference/TraitA-class.md)
  : Additive trait
- [`TraitA2-class`](https://gaynorr.github.io/AlphaSimR/reference/TraitA2-class.md)
  : Sex specific additive trait
- [`TraitA2D-class`](https://gaynorr.github.io/AlphaSimR/reference/TraitA2D-class.md)
  : Sex specific additive and dominance trait
- [`TraitAD-class`](https://gaynorr.github.io/AlphaSimR/reference/TraitAD-class.md)
  : Additive and dominance trait
- [`TraitADE-class`](https://gaynorr.github.io/AlphaSimR/reference/TraitADE-class.md)
  : Additive, dominance, and epistatic trait
- [`TraitADEG-class`](https://gaynorr.github.io/AlphaSimR/reference/TraitADEG-class.md)
  : Additive, dominance, epistasis, and GxE trait
- [`TraitADG-class`](https://gaynorr.github.io/AlphaSimR/reference/TraitADG-class.md)
  : Additive, dominance and GxE trait
- [`TraitAE-class`](https://gaynorr.github.io/AlphaSimR/reference/TraitAE-class.md)
  : Additive and epistatic trait
- [`TraitAEG-class`](https://gaynorr.github.io/AlphaSimR/reference/TraitAEG-class.md)
  : Additive, epistasis and GxE trait
- [`TraitAG-class`](https://gaynorr.github.io/AlphaSimR/reference/TraitAG-class.md)
  : Additive and GxE trait
- [`aa()`](https://gaynorr.github.io/AlphaSimR/reference/aa.md) :
  Additive-by-additive epistatic deviations
- [`addSegSite()`](https://gaynorr.github.io/AlphaSimR/reference/addSegSite.md)
  : Add segregating site to MapPop
- [`asCategorical()`](https://gaynorr.github.io/AlphaSimR/reference/asCategorical.md)
  : Convert a normal (Gaussian) trait to an ordered categorical
  (threshold) trait
- [`attrition()`](https://gaynorr.github.io/AlphaSimR/reference/attrition.md)
  : Lose individuals at random
- [`bv()`](https://gaynorr.github.io/AlphaSimR/reference/bv.md) :
  Breeding value
- [`cChr()`](https://gaynorr.github.io/AlphaSimR/reference/cChr.md) :
  Combine MapPop chromosomes
- [`calcGCA()`](https://gaynorr.github.io/AlphaSimR/reference/calcGCA.md)
  : Calculate GCA
- [`dd()`](https://gaynorr.github.io/AlphaSimR/reference/dd.md) :
  Dominance deviations
- [`doubleGenome()`](https://gaynorr.github.io/AlphaSimR/reference/doubleGenome.md)
  : Double the ploidy of individuals
- [`ebv()`](https://gaynorr.github.io/AlphaSimR/reference/ebv.md) :
  Estimated breeding value
- [`editGenome()`](https://gaynorr.github.io/AlphaSimR/reference/editGenome.md)
  : Edit genome
- [`editGenomeTopQtl()`](https://gaynorr.github.io/AlphaSimR/reference/editGenomeTopQtl.md)
  : Edit genome - the top QTL
- [`fastRRBLUP()`](https://gaynorr.github.io/AlphaSimR/reference/fastRRBLUP.md)
  : Fast RR-BLUP
- [`genParam()`](https://gaynorr.github.io/AlphaSimR/reference/genParam.md)
  : Sumarize genetic parameters
- [`genicVarA()`](https://gaynorr.github.io/AlphaSimR/reference/genicVarA.md)
  : Additive genic variance
- [`genicVarAA()`](https://gaynorr.github.io/AlphaSimR/reference/genicVarAA.md)
  : Additive-by-additive genic variance
- [`genicVarD()`](https://gaynorr.github.io/AlphaSimR/reference/genicVarD.md)
  : Dominance genic variance
- [`genicVarG()`](https://gaynorr.github.io/AlphaSimR/reference/genicVarG.md)
  : Total genic variance
- [`getGenMap()`](https://gaynorr.github.io/AlphaSimR/reference/getGenMap.md)
  : Get genetic map
- [`getNumThreads()`](https://gaynorr.github.io/AlphaSimR/reference/getNumThreads.md)
  : Number of available threads
- [`getPed()`](https://gaynorr.github.io/AlphaSimR/reference/getPed.md)
  : Get pedigree
- [`getQtlMap()`](https://gaynorr.github.io/AlphaSimR/reference/getQtlMap.md)
  : Get QTL genetic map
- [`getSnpMap()`](https://gaynorr.github.io/AlphaSimR/reference/getSnpMap.md)
  : Get SNP genetic map
- [`gv()`](https://gaynorr.github.io/AlphaSimR/reference/gv.md) :
  Genetic value
- [`hybridCross()`](https://gaynorr.github.io/AlphaSimR/reference/hybridCross.md)
  : Hybrid crossing
- [`importGenMap()`](https://gaynorr.github.io/AlphaSimR/reference/importGenMap.md)
  : Import genetic map
- [`importHaplo()`](https://gaynorr.github.io/AlphaSimR/reference/importHaplo.md)
  : Import haplotypes
- [`importInbredGeno()`](https://gaynorr.github.io/AlphaSimR/reference/importInbredGeno.md)
  : Import inbred, diploid genotypes
- [`isFemale()`](https://gaynorr.github.io/AlphaSimR/reference/isFemale.md)
  [`isMale()`](https://gaynorr.github.io/AlphaSimR/reference/isFemale.md)
  : Test if individuals of a population are female or male
- [`isPop()`](https://gaynorr.github.io/AlphaSimR/reference/isPop.md) :
  Test if object is of a Population class
- [`makeCross()`](https://gaynorr.github.io/AlphaSimR/reference/makeCross.md)
  : Make designed crosses
- [`makeCross2()`](https://gaynorr.github.io/AlphaSimR/reference/makeCross2.md)
  : Make designed crosses
- [`makeDH()`](https://gaynorr.github.io/AlphaSimR/reference/makeDH.md)
  : Generates DH lines
- [`meanEBV()`](https://gaynorr.github.io/AlphaSimR/reference/meanEBV.md)
  : Mean estimated breeding values
- [`meanG()`](https://gaynorr.github.io/AlphaSimR/reference/meanG.md) :
  Mean genetic values
- [`meanP()`](https://gaynorr.github.io/AlphaSimR/reference/meanP.md) :
  Mean phenotypic values
- [`mendelianSampling()`](https://gaynorr.github.io/AlphaSimR/reference/mendelianSampling.md)
  : Calculate Mendelian sampling
- [`mergeGenome()`](https://gaynorr.github.io/AlphaSimR/reference/mergeGenome.md)
  : Combine genomes of individuals
- [`mergePops()`](https://gaynorr.github.io/AlphaSimR/reference/mergePops.md)
  : Merge list of populations
- [`mutate()`](https://gaynorr.github.io/AlphaSimR/reference/mutate.md)
  : Add Random Mutations
- [`nInd()`](https://gaynorr.github.io/AlphaSimR/reference/nInd.md) :
  Number of individuals
- [`newEmptyPop()`](https://gaynorr.github.io/AlphaSimR/reference/newEmptyPop.md)
  : Creates an empty population
- [`newMapPop()`](https://gaynorr.github.io/AlphaSimR/reference/newMapPop.md)
  : New MapPop
- [`newMultiPop()`](https://gaynorr.github.io/AlphaSimR/reference/newMultiPop.md)
  : Create new Multi Population
- [`newPop()`](https://gaynorr.github.io/AlphaSimR/reference/newPop.md)
  : Create new population
- [`parentAverage()`](https://gaynorr.github.io/AlphaSimR/reference/parentAverage.md)
  : Calculate parent average
- [`pedigreeCross()`](https://gaynorr.github.io/AlphaSimR/reference/pedigreeCross.md)
  : Pedigree cross
- [`pheno()`](https://gaynorr.github.io/AlphaSimR/reference/pheno.md) :
  Phenotype
- [`popVar()`](https://gaynorr.github.io/AlphaSimR/reference/popVar.md)
  : Population variance
- [`pullIbdHaplo()`](https://gaynorr.github.io/AlphaSimR/reference/pullIbdHaplo.md)
  : Pull IBD haplotypes
- [`pullMarkerGeno()`](https://gaynorr.github.io/AlphaSimR/reference/pullMarkerGeno.md)
  : Pull marker genotypes
- [`pullMarkerHaplo()`](https://gaynorr.github.io/AlphaSimR/reference/pullMarkerHaplo.md)
  : Pull marker haplotypes
- [`pullQtlGeno()`](https://gaynorr.github.io/AlphaSimR/reference/pullQtlGeno.md)
  : Pull QTL genotypes
- [`pullQtlHaplo()`](https://gaynorr.github.io/AlphaSimR/reference/pullQtlHaplo.md)
  : Pull QTL haplotypes
- [`pullSegSiteGeno()`](https://gaynorr.github.io/AlphaSimR/reference/pullSegSiteGeno.md)
  : Pull segregating site genotypes
- [`pullSegSiteHaplo()`](https://gaynorr.github.io/AlphaSimR/reference/pullSegSiteHaplo.md)
  : Pull seg site haplotypes
- [`pullSnpGeno()`](https://gaynorr.github.io/AlphaSimR/reference/pullSnpGeno.md)
  : Pull SNP genotypes
- [`pullSnpHaplo()`](https://gaynorr.github.io/AlphaSimR/reference/pullSnpHaplo.md)
  : Pull SNP haplotypes
- [`quickHaplo()`](https://gaynorr.github.io/AlphaSimR/reference/quickHaplo.md)
  : Quick founder haplotype simulation
- [`randCross()`](https://gaynorr.github.io/AlphaSimR/reference/randCross.md)
  : Make random crosses
- [`randCross2()`](https://gaynorr.github.io/AlphaSimR/reference/randCross2.md)
  : Make random crosses
- [`reduceGenome()`](https://gaynorr.github.io/AlphaSimR/reference/reduceGenome.md)
  : Create individuals with reduced ploidy
- [`resetPop()`](https://gaynorr.github.io/AlphaSimR/reference/resetPop.md)
  : Reset population
- [`runMacs()`](https://gaynorr.github.io/AlphaSimR/reference/runMacs.md)
  : Create founder haplotypes using MaCS
- [`runMacs2()`](https://gaynorr.github.io/AlphaSimR/reference/runMacs2.md)
  : Alternative wrapper for MaCS
- [`sampleHaplo()`](https://gaynorr.github.io/AlphaSimR/reference/sampleHaplo.md)
  : Sample haplotypes from a MapPop
- [`selIndex()`](https://gaynorr.github.io/AlphaSimR/reference/selIndex.md)
  : Selection index
- [`selInt()`](https://gaynorr.github.io/AlphaSimR/reference/selInt.md)
  : Selection intensity
- [`selectCross()`](https://gaynorr.github.io/AlphaSimR/reference/selectCross.md)
  : Select and randomly cross
- [`selectFam()`](https://gaynorr.github.io/AlphaSimR/reference/selectFam.md)
  : Select families
- [`selectInd()`](https://gaynorr.github.io/AlphaSimR/reference/selectInd.md)
  : Select individuals
- [`selectOP()`](https://gaynorr.github.io/AlphaSimR/reference/selectOP.md)
  : Select open pollinating plants
- [`selectWithinFam()`](https://gaynorr.github.io/AlphaSimR/reference/selectWithinFam.md)
  : Select individuals within families
- [`self()`](https://gaynorr.github.io/AlphaSimR/reference/self.md) :
  Self individuals
- [`setEBV()`](https://gaynorr.github.io/AlphaSimR/reference/setEBV.md)
  : Set estimated breeding values (EBV)
- [`setMarkerHaplo()`](https://gaynorr.github.io/AlphaSimR/reference/setMarkerHaplo.md)
  : Set marker haplotypes
- [`setPheno()`](https://gaynorr.github.io/AlphaSimR/reference/setPheno.md)
  : Set phenotypes
- [`setPhenoGCA()`](https://gaynorr.github.io/AlphaSimR/reference/setPhenoGCA.md)
  : Set GCA as phenotype
- [`setPhenoProgTest()`](https://gaynorr.github.io/AlphaSimR/reference/setPhenoProgTest.md)
  : Set progeny test as phenotype
- [`smithHazel()`](https://gaynorr.github.io/AlphaSimR/reference/smithHazel.md)
  : Calculate Smith-Hazel weights
- [`solveMKM()`](https://gaynorr.github.io/AlphaSimR/reference/solveMKM.md)
  : Solve Multikernel Model
- [`solveMVM()`](https://gaynorr.github.io/AlphaSimR/reference/solveMVM.md)
  : Solve Multivariate Model
- [`solveRRBLUP()`](https://gaynorr.github.io/AlphaSimR/reference/solveRRBLUP.md)
  : Solve RR-BLUP
- [`solveRRBLUPMK()`](https://gaynorr.github.io/AlphaSimR/reference/solveRRBLUPMK.md)
  : Solve Multikernel RR-BLUP
- [`solveRRBLUPMV()`](https://gaynorr.github.io/AlphaSimR/reference/solveRRBLUPMV.md)
  : Solve Multivariate RR-BLUP
- [`solveRRBLUP_EM()`](https://gaynorr.github.io/AlphaSimR/reference/solveRRBLUP_EM.md)
  : Solve RR-BLUP with EM
- [`solveRRBLUP_EM2()`](https://gaynorr.github.io/AlphaSimR/reference/solveRRBLUP_EM2.md)
  : Solve RR-BLUP with EM and 2 random effects
- [`solveRRBLUP_EM3()`](https://gaynorr.github.io/AlphaSimR/reference/solveRRBLUP_EM3.md)
  : Solve RR-BLUP with EM and 3 random effects
- [`solveUVM()`](https://gaynorr.github.io/AlphaSimR/reference/solveUVM.md)
  : Solve Univariate Model
- [`usefulness()`](https://gaynorr.github.io/AlphaSimR/reference/usefulness.md)
  : Usefulness criterion
- [`varA()`](https://gaynorr.github.io/AlphaSimR/reference/varA.md) :
  Additive variance
- [`varAA()`](https://gaynorr.github.io/AlphaSimR/reference/varAA.md) :
  Additive-by-additive epistatic variance
- [`varD()`](https://gaynorr.github.io/AlphaSimR/reference/varD.md) :
  Dominance variance
- [`varEBV()`](https://gaynorr.github.io/AlphaSimR/reference/varEBV.md)
  : Variance of estimated breeding values
- [`varG()`](https://gaynorr.github.io/AlphaSimR/reference/varG.md) :
  Total genetic variance
- [`varP()`](https://gaynorr.github.io/AlphaSimR/reference/varP.md) :
  Phenotypic variance
- [`writePlink()`](https://gaynorr.github.io/AlphaSimR/reference/writePlink.md)
  : Writes a Pop-class as PLINK files
- [`writeRecords()`](https://gaynorr.github.io/AlphaSimR/reference/writeRecords.md)
  : Write data records
