<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Traits in AlphaSimR • AlphaSimR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Traits in AlphaSimR">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">AlphaSimR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.6.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/AlphaSimR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/DataImport.html">Data Import</a></li>
    <li><a class="dropdown-item" href="../articles/Genome.html">Genome</a></li>
    <li><a class="dropdown-item" href="../articles/GenomicSelection.html">Genomic Selection</a></li>
    <li><a class="dropdown-item" href="../articles/intro.html">Introduction to AlphaSimR</a></li>
    <li><a class="dropdown-item" href="../articles/Meiosis.html">Meiosis</a></li>
    <li><a class="dropdown-item" href="../articles/Misc.html">Misc</a></li>
    <li><a class="dropdown-item" href="../articles/QuantGen.html">Quantitative Genetics</a></li>
    <li><a class="dropdown-item" href="../articles/traits.html">Traits in AlphaSimR</a></li>
    <li><a class="dropdown-item" href="../articles/Traits.html">Traits</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/gaynorr/AlphaSimR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Traits in AlphaSimR</h1>
                        <h4 data-toc-skip class="author">Chris
Gaynor</h4>
            
            <h4 data-toc-skip class="date">2024-11-01</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/gaynorr/AlphaSimR/blob/master/vignettes/traits.Rmd" class="external-link"><code>vignettes/traits.Rmd</code></a></small>
      <div class="d-none name"><code>traits.Rmd</code></div>
    </div>

    
    
<p>This vignette describes AlphaSimR’s biological model for traits. The
biological model is responsible for converting an individual’s genotype
into a genetic value. The genetic value is used to create a phenotype.
AlphaSimR’s biological model is primarily based on classic models used
in quantitative genetics. Thus, users that have taken an introductory
quantitative genetics course should already be familiar with most
elements of AlphaSimR’s biological model. This vignette will assume the
reader is such a person and thus focus on aspects of AlphaSimR’s
biological model that may not be obvious or don’t follow classic models.
The vignette will begin by describing a novel classification system used
in AlphaSimR to name different types of traits. Then it will present the
full biological model along with a detailed description for each
component in the model.</p>
<p>Traits in AlphaSimR are classified according to the biological
effects they model using the <strong>ADEG</strong> framework. Under this
framework, each trait is assigned a name consisting of one or more
letters. The letters come from the name <strong>ADEG</strong>, whose
letters correspond to biological effects. The biological effects are:
<strong>A</strong>dditive, <strong>D</strong>ominance,
<strong>E</strong>pistatic and <strong>G</strong>enotype-by-environment.
For example, a trait with only additive effects is called an
<strong>A</strong> trait and a trait with both additive and dominance
effects is called an <strong>AD</strong> trait. The following traits are
modeled in AlphaSimR: <strong>A</strong>, <strong>AD</strong>,
<strong>AE</strong>, <strong>AG</strong>, <strong>ADE</strong>,
<strong>ADG</strong>, <strong>AEG</strong>, and
<strong>ADEG</strong>.</p>
<p>The most complex trait in AlphaSimR is an <strong>ADEG</strong>
trait, because it includes all biological effects. This trait is
represented using the equation below.</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>G</mi><mi>V</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo>,</mo><mi>w</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mi>μ</mi><mo>+</mo><mi>A</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><mi>D</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><mi>E</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><mi>G</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo>,</mo><mi>w</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\begin{equation}
GV(x, w) = 
\mu + 
A(x) +
D(x) +
E(x) +
G(x, w)
\end{equation}</annotation></semantics></math></p>
<p>The left-hand side of equation (1) consists of the following terms:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>G</mi><mi>V</mi></mrow><annotation encoding="application/x-tex">GV</annotation></semantics></math>,
which represents an individual’s genetic value;
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>x</mi><annotation encoding="application/x-tex">x</annotation></semantics></math>,
which represents a vector of QTL genotype dosages; and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>w</mi><annotation encoding="application/x-tex">w</annotation></semantics></math>,
which represents an environmental covariate. The right-hand side of the
equation consists of an intercept
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>μ</mi><annotation encoding="application/x-tex">\mu</annotation></semantics></math>)
and four functions. The intercept is used to obtain the user specified
trait mean in a founder population. The functions model each of the
biological effects. All other trait in the <strong>ADEG</strong>
framework can be modeled with equation (1) by simply removing any
functions for biological effects that those traits lack.</p>
<p>The discussion of the <strong>ADEG</strong> framework will continue
below with an explanation for each function in equation (1) after first
describing the concept of genotype dosage scaling. Genotype dosage
scaling is discussed first, because it is applied within each of the
aforementioned functions.</p>
<div class="section level2">
<h2 id="genotype-dosage-scaling">Genotype Dosage Scaling<a class="anchor" aria-label="anchor" href="#genotype-dosage-scaling"></a>
</h2>
<p>AlphaSimR defines an individual’s raw genotype dosage as the number
of copies of the “1” allele at a locus. Since all loci in AlphaSimR are
biallelic with alleles “0” and “1”, this definition for raw genotype
dosage fully explains an individual’s genotype. The number of copies of
the “0” allele is just the individual’s ploidy level minus its raw
genotype dosage. This is usually a rather irrelevant detail, because
most user will want to model diploid organism whose ploidy level is
always two. However, AlphaSimR can also be used to model a wide range of
autopolyploid organisms and it even allows for mixing ploidy levels
within a simulation. This means that the allowable range for raw
genotype dosage is not a fixed value and that software must account for
this fact.</p>
<p>AlphaSimR accounts for different levels of ploidy by scaling an
organism’s genotype dosage in accordance with its ploidy level. The
primary motivation for using scaled dosages is to unify user inputs. The
use of scaled dosages can also make it easier to compare simulations
with different levels of ploidy. However, the user must use care when
making such comparisons, because the underlying assumptions used by this
model may not be valid.</p>
<p>The use of scaled genotype dosages implies that an individual’s
genetic value depends on the relative ratio of alleles and that it is
independent of the organism’s ploidy level. For some traits in some
organisms there is evidence to support this assumption or at least not
reject it, but there is also plenty of evidence clearly rejecting this
assumption in other traits <span class="citation">(Gallais 2003)</span>.
Thus, the direct comparisons that are possible in AlphaSimR are not
always biologically relevant. However, the vast majority of users won’t
be running a simulation that depends on this assumption, so the
information presented below is solely for the purpose of understanding
the functions presented in the following sections.</p>
<p>There are two types of scaled genotype dosages in AlphaSimR: additive
and dominance. An explanation for both is given below. This is followed
by a table providing an example of dosage scaling in both diploid and
autotetraploid organisms.</p>
<p>The scaled additive genotype dosage
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>x</mi><mi>A</mi></msub><annotation encoding="application/x-tex">x_A</annotation></semantics></math>)
is shown in equation (2). This equation linearly scales relative dosage
to set the values for opposing homozygotes to -1 and 1.</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>x</mi><mi>A</mi></msub><mo>=</mo><mo minsize="1.2" maxsize="1.2" stretchy="false" form="prefix">(</mo><mi>x</mi><mo>−</mo><mfrac displaystyle="false"><mrow><mi>p</mi><mi>l</mi><mi>o</mi><mi>i</mi><mi>d</mi><mi>y</mi></mrow><mn>2</mn></mfrac><mo minsize="1.2" maxsize="1.2" stretchy="false" form="postfix">)</mo><mo minsize="1.2" maxsize="1.2" stretchy="false" form="prefix">(</mo><mfrac displaystyle="false"><mn>2</mn><mrow><mi>p</mi><mi>l</mi><mi>o</mi><mi>i</mi><mi>d</mi><mi>y</mi></mrow></mfrac><mo minsize="1.2" maxsize="1.2" stretchy="false" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">\begin{equation}
x_A = \big( x - \tfrac{ploidy}{2} \big) \big( \tfrac{2}{ploidy} \big)
\end{equation}</annotation></semantics></math></p>
<p>The scaled dominance genotype dosage
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>x</mi><mi>D</mi></msub><annotation encoding="application/x-tex">x_D</annotation></semantics></math>)
is shown in equation (3). This equation uses non-linear scaling to fit
the value of the opposing homozygotes to 0 and middlemost heterozygote
to 1. The middlemost heterozygote is the genotype with an equal ratio of
“0” and “1” alleles. For a diploid organism, the scaled dominance
genotype dosage matches the classic parameterization for dominance. For
autopolyploid organisms, the scaled dominance genotype dosage is
consistent with digenic dominance (discussed later).</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>x</mi><mi>D</mi></msub><mo>=</mo><mi>x</mi><mo minsize="1.2" maxsize="1.2" stretchy="false" form="prefix">(</mo><mi>p</mi><mi>l</mi><mi>o</mi><mi>i</mi><mi>d</mi><mi>y</mi><mo>−</mo><mi>x</mi><mo minsize="1.2" maxsize="1.2" stretchy="false" form="postfix">)</mo><mo minsize="1.2" maxsize="1.2" stretchy="false" form="prefix">(</mo><mfrac displaystyle="false"><mn>2</mn><mrow><mi>p</mi><mi>l</mi><mi>o</mi><mi>i</mi><mi>d</mi><mi>y</mi></mrow></mfrac><msup><mo minsize="1.2" maxsize="1.2" stretchy="false" form="postfix">)</mo><mn>2</mn></msup></mrow><annotation encoding="application/x-tex">\begin{equation}
x_D = x \big( ploidy - x \big) \big( \tfrac{2}{ploidy} \big)^2
\end{equation}</annotation></semantics></math></p>
<p>Table 1 provides an example of raw and scaled genotype dosages for a
diploid and an autotetraploid organism. The diploid and tetraploid
columns represent the raw genotype dosages for the respective organisms.
The additive and dominance columns represent the corresponding scaled
genotype dosages. This table shows how the 0, 1 and 2 genotypes of a
diploid organism are treated as being equivalent to the 0, 2 and 4
genotypes of an autotetraploid organism.</p>
</div>
<div class="section level2">
<h2 id="additive-effects">Additive Effects<a class="anchor" aria-label="anchor" href="#additive-effects"></a>
</h2>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mo>∑</mo><mi>a</mi><msub><mi>x</mi><mi>A</mi></msub></mrow><annotation encoding="application/x-tex">\begin{equation}
A(x) = \sum a x_A
\end{equation}</annotation></semantics></math></p>
<p>The function for additive effects is given above in equation (4). The
right-hand side is a summation over all QTL for the product of the
additive effect
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>a</mi><annotation encoding="application/x-tex">a</annotation></semantics></math>)
and the scaled additive dosage
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>x</mi><mi>A</mi></msub><annotation encoding="application/x-tex">x_A</annotation></semantics></math>).
This equation is equivalent to parameterizations in classic quantitative
trait models. The only unique aspect of additive effects in AlphaSimR is
the sampling of those effects.</p>
<p>Additive effects are sampled in two stages. The first stage involves
sampling initial values and is similar to methods used by other
stochastic simulation software programs. It is the second stage that is
unique. This stage involves scaling the magnitude of the initial values
to achieve a desired genetic variance, that being either total or
additive genetic variance.</p>
<p>The first stage of sampling additive effects is setting initial
values for the effects. Initial values are sampled from either a
standard normal distribution or a gamma distribution. When using a gamma
distribution, the user specifies the shape parameter and the scale
parameter is set to 1. The deviates sampled from a gamma distribution
are randomly assigned either a positive or negative sign. Random
assignment of this sign results in an expected distribution that is
symmetric distribution with mean 0.</p>
<p>The second stage is scaling the magnitude of the effects to achieve a
user specified genetic variance. The user specified genetic variance can
be either total or additive. The scaling procedure involves not just the
additive effect, but also dominance and epistatic effects if the trait
includes those effects as well. The procedure works by first calculating
the variance in the founder population using the initially sampled
effects and then calculating a scaling constant that is applied to all
effects to achieve the desired variance in the founder population. The
scaling constant equals the square-root of the target genetic variance
(total or additive) divided by the square-root of the initial genetic
variance (total or additive). Note that the founder population is the
population the user uses to initialize the simulation parameters
object.</p>
<p>Scaling the magnitude of effects allows AlphaSimR to build
simulations with trait values matching real-world counterparts. For
example, a user simulating grain yield in a plant species is likely to
have estimates for the means and variances in tons per hectare. The user
can use those to create a simulation with the same values. The primary
benefit of matching these values is to make interpretation of simulation
results more intuitive. The significance of the results will be more
apparent to the user because they can work in a scale that is familiar
to them. It also allows the user to more easily identify potential flaws
in the simulation when values move outside the range of biologically
acceptable values.</p>
</div>
<div class="section level2">
<h2 id="dominance-effects">Dominance Effects<a class="anchor" aria-label="anchor" href="#dominance-effects"></a>
</h2>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>D</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mo>∑</mo><mi>d</mi><msub><mi>x</mi><mi>D</mi></msub></mrow><annotation encoding="application/x-tex">\begin{equation}
D(x) = \sum d x_D
\end{equation}</annotation></semantics></math></p>
<p>The function for dominance effects is given above in equation (5).
The right-hand side of the equation is a summation over all QTL for the
product of the dominance effect
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>d</mi><annotation encoding="application/x-tex">d</annotation></semantics></math>)
and the scaled dominance dosage
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>x</mi><mi>D</mi></msub><annotation encoding="application/x-tex">x_D</annotation></semantics></math>).
This equation is equivalent to the parameterization of dominance in
classic quantitative trait models for diploid organisms. As with the
additive effects, the method for sampling dominance effects in AlphaSimR
requires special attention.</p>
<p>Dominance effects are calculated in AlphaSimR using the concept of
dominance degrees. The formula for this calculation is given below in
equation (6). The equation shows that the dominance effect
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>d</mi><annotation encoding="application/x-tex">d</annotation></semantics></math>)
at a locus is the dominance degree
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>δ</mi><annotation encoding="application/x-tex">\delta</annotation></semantics></math>)
at that locus times the absolute value of its additive effect
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>a</mi><annotation encoding="application/x-tex">a</annotation></semantics></math>).</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>d</mi><mo>=</mo><mi>δ</mi><mrow><mo stretchy="true" form="prefix">|</mo><mi>a</mi><mo stretchy="true" form="postfix">|</mo></mrow></mrow><annotation encoding="application/x-tex">\begin{equation}
d = \delta \left| a \right|
\end{equation}</annotation></semantics></math></p>
<p>The rationale behind using dominance degrees is their intuitive
biological interpretation in diploid organisms. For example, a dominance
degree of 0 represents no dominance and an additive model. A dominance
degree of 1 corresponds to complete dominance. Dominance degrees between
0 and 1 correspond to partial dominance, and values above 1 correspond
to over-dominance.</p>
<p>Dominance effects, as with additive effects, are sampled in two
stages. The first stage is sampling of initial values and the second
stage is scaling the magnitude of those values. The sampling of initial
effects involves two user supplied parameters, the mean and variance for
a normal distribution used to sample dominance degrees. The dominance
degrees are then used in conjunction with the additive effects to
calculate dominance effects. The scaling procedure is then performed as
described above in the additive effects section to calculate the scaling
constant. The scaling constant is then directly applied to the dominance
effects. Note that this scaling changes the value of the dominance
effect, but does not change the value of the dominance degree. This
means that the specification of the dominance degree distribution is
independent of the requested genetic variance.</p>
<p>Dominance effects become more complicated when discussing polyploid
organisms. This is because polyploids have additional heterozygous
genotypes. For example, an autotetraploid organism has two homozygous
genotypes (0 and 4) and three heterozygous genotypes (1, 2 and 3). Each
additional heterozygous genotype requires an additional dominance
parameter to obtain a fully parameterized model. Several different
parameterizations have been used for polyploids and a full discussion of
these parameterizations is outside the scope of this document.
Interested readers should refer to Gallais’s textbook for more details
<span class="citation">(2003)</span>.</p>
<p>AlphaSimR does not include additional term to its dominance model for
polyploids, so it does not make use of a fully parameterized model.
Instead, AlphaSimR uses a digenic dominance model for all ploidy levels
due to its use of scaled dominance genotype dosages. This model was
chosen because it provides consistency in user interface and internal
coding regardless of ploidy level and it provides as a reasonable
approximation for partial dominance in highly polygenic traits.</p>
<p>An unfortunate side effect of the dominance model in polyploids is
that the previously described interpretation of dominance degrees
breaks. Consider a single QTL with an additive effect of 1 and a
dominance degree of 1. The genotypes and genetic values for the example
for both a diploid and an autotetraploid organism are given below in
Table 2. For the diploid organism, the heterozygous genotype is
equivalent to the best homozygote, so a dominance degree of 1 indicates
complete dominance. However, the autotetraploid organism has three
heterozygous genotypes. The value of the middlemost heterozygous
genotype is equivalent to the best homozygote, but it is not the
heterozygous genotype with the highest value. The heterozygous genotype
with the highest value is better than the best homozygote, so a
dominance degree of 1 actually represents over-dominance in an
autotetraploid.</p>
</div>
<div class="section level2">
<h2 id="epistatic-effects">Epistatic Effects<a class="anchor" aria-label="anchor" href="#epistatic-effects"></a>
</h2>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mo>∑</mo><mi>e</mi><msub><mi>x</mi><msub><mi>A</mi><mn>1</mn></msub></msub><msub><mi>x</mi><msub><mi>A</mi><mn>2</mn></msub></msub></mrow><annotation encoding="application/x-tex">\begin{equation}
E(x) = \sum e x_{A_1} x_{A_2}
\end{equation}</annotation></semantics></math></p>
<p>The function for epistatic effects is given above in equation (7).
The summation in the right-hand side is over pairs of QTLs, because
AlphaSimR uses a simplified model of epistasis that restricts epistatic
interactions to pairs of loci. Each QTL must be present in one and only
one pair, so the number of pairs is equal to half the number of QTL. The
remaining elements in the right-hand side of equation (7) are epistatic
effects
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>e</mi><annotation encoding="application/x-tex">e</annotation></semantics></math>),
the scaled additive dosage for the first locus in a pair
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>x</mi><msub><mi>A</mi><mn>1</mn></msub></msub><annotation encoding="application/x-tex">x_{A_1}</annotation></semantics></math>)
and the scaled additive dosage for the second locus in a pair
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>x</mi><msub><mi>A</mi><mn>2</mn></msub></msub><annotation encoding="application/x-tex">x_{A_2}</annotation></semantics></math>).</p>
<p>The epistatic model in equation (7) is somewhat constrained. The
model only allows epistasis between pairs of loci and it only models
additive-by-additive epistasis. The motivation behind using this
constrained model is to maintain computational tractability when using
very large numbers of QTL. As with the dominance model, this model is
considered a reasonable approximation for a more complicated
reality.</p>
<p>The sampling of epistatic effects is similar to the sampling of
additive effects with one additional user specified parameter for the
relative ratio of the additive-by-additive epistatic variance. The
relative ratio refers to the ratio between additive-by-additive
epistatic variance and additive variance. This parameter sets the
variance of epistatic effects. Specifically, the variance of the effects
is set to a value expected to achieve the desired ratio in a random
mating population whose QTL are in linkage equilibrium and have an
allele frequency of 0.5. Note that most populations, including simulated
populations, will not meet these assumptions, so the observed relative
ratio between additive and additive-by-additive epistatic variance is
unlikely to match the requested ratio. The observed additive-by-additive
epistatic variance will typically be smaller, because its value is
maximized at allele frequency 0.5. Note that unselected bi-parental
populations derived from inbred parents are expected to have allele
frequencies of 0.5 for segregating alleles. Thus, these populations are
ideal for estimating a reasonable value for the relative ratio of
additive-by-additive epistatic variance.</p>
</div>
<div class="section level2">
<h2 id="genotype-by-environment-effects">Genotype-by-Environment Effects<a class="anchor" aria-label="anchor" href="#genotype-by-environment-effects"></a>
</h2>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>G</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo>,</mo><mi>w</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mi>w</mi><mi>b</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\begin{equation}
G(x, w) = w b(x)
\end{equation}</annotation></semantics></math></p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>b</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msub><mi>μ</mi><mi>G</mi></msub><mo>+</mo><mo>∑</mo><mi>g</mi><msub><mi>x</mi><mi>A</mi></msub></mrow><annotation encoding="application/x-tex">\begin{equation}
b(x) = \mu_G + \sum g x_A
\end{equation}</annotation></semantics></math></p>
<p>The function for genotype-by-environment effects is given above in
equations (8) and (9). The right-hand side of the equation (8) contains
two parts: an environmental covariate
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>w</mi><annotation encoding="application/x-tex">w</annotation></semantics></math>)
and a genotype specific slope
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>b</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>x</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">b(x)</annotation></semantics></math>).
The formula for the genotype specific slope is shown in equation (9).
The right-hand side of this equation includes an intercept value
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>μ</mi><mi>G</mi></msub><annotation encoding="application/x-tex">\mu_G</annotation></semantics></math>)
and a summation over all QTL for the product of a
genotype-by-environment effect
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>)
and the scaled additive dosage
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>x</mi><mi>A</mi></msub><annotation encoding="application/x-tex">x_A</annotation></semantics></math>).</p>
<p>The following paragraphs explain how the above equations model
genotype-by-environment interactions. Initially, the explanation will
cover a case where a population has genotype-by-environment interaction
variance, but no environmental variance. An explanation for how the
parameters of these equations are changed to model both
genotype-by-environment interaction variance and environmental variance
is given in the next section.</p>
<p>To begin, a description of how the variables in the above equations
are sampled is needed. This is because it is much easier to understand
how the formulas come together to model genotype-by-environment
interactions when it is understood what variables represent.</p>
<p>The environmental covariate
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>w</mi><annotation encoding="application/x-tex">w</annotation></semantics></math>)
in equation (8) represents an environmental component of the
genotype-by-environment interaction. The value of the environmental
covariate is randomly sampled from a standard normal distribution. By
definition, this means that the average value of the environmental
covariate is zero and its variance is one. The average value of the
environmental covariate is considered to be the target environment.
Thus, the value for equation (8) in the target environment is always
zero.</p>
<p>The genotype specific slope in equation (9) represents the genetic
component of the genotype-by-environment interaction. The astute reader
will notice that this equation is very similar to the equation for an
additive trait. Indeed, the genotype specific slope is really just an
additive trait. The term
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>μ</mi><mi>G</mi></msub><annotation encoding="application/x-tex">\mu_G</annotation></semantics></math>
serves a similar role to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>μ</mi><annotation encoding="application/x-tex">\mu</annotation></semantics></math>
in equation (1) and the summation on the right-hand side of equation (9)
is similar to the function for additive effects in equation (4). Also
like an additive trait, effects for the genotype specific slope are
scaled to achieve a specific mean and variance in the founder
population. In this case, the mean is set to zero and the variance is
set to the user specified genotype-by-environment interaction
variance.</p>
<p>To understand how the environmental covariate and the genotype
specific slope model genotype-by-environment interactions it helps to
review the properties of these variables. The environmental covariate is
just a random variable with a mean of zero and a variance of one. The
genotype specific slope is also random variable, with regards to the
founder population, that has a mean of zero and a variance equal to the
genotype-by-environment variance. This means that equation (8) is just
the product of two random variables. These random variables are
independent, so the formula for the variance of this product is given in
the equation below.</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>V</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>w</mi><mi>b</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mi>E</mi><msup><mrow><mo stretchy="true" form="prefix">[</mo><mi>w</mi><mo stretchy="true" form="postfix">]</mo></mrow><mn>2</mn></msup><mi>V</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>b</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><mi>V</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>w</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>E</mi><msup><mrow><mo stretchy="true" form="prefix">[</mo><mi>b</mi><mo stretchy="true" form="postfix">]</mo></mrow><mn>2</mn></msup><mo>+</mo><mi>V</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>w</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>V</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>b</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\begin{equation}
Var(w b) = E[w]^2 Var(b) + Var(w) E[b]^2 + Var(w) Var(b)
\end{equation}</annotation></semantics></math></p>
<p>Equation (10) above gives the variance for the product of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>w</mi><annotation encoding="application/x-tex">w</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>b</mi><annotation encoding="application/x-tex">b</annotation></semantics></math>
in equation (8). The term
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>V</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Var()</annotation></semantics></math>
indicates the variance of a variable and the term
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mrow><mo stretchy="true" form="prefix">[</mo><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">E[]</annotation></semantics></math>
indicates the expectation (mean) of a variable. In the founder
population, the expectations for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>w</mi><annotation encoding="application/x-tex">w</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>b</mi><annotation encoding="application/x-tex">b</annotation></semantics></math>
are zero, so the first two terms in the right-hand side of the equation
drop out. Equation (10) reduces to the product of the two variances. The
variance of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>w</mi><annotation encoding="application/x-tex">w</annotation></semantics></math>
is one and the variance of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>b</mi><annotation encoding="application/x-tex">b</annotation></semantics></math>
equals the genotype-by-environment interaction variance, so the variance
in equation (8) equals the genotype-by-environment interaction variance
and is equivalent to the variance in the genotype specific slopes. It
must be repeated that the above description is limited to the founder
population. This is an important point, because the mean and variance
for genotype specific slope can and will be different in other
populations since it is under genetic control.</p>
<div class="section level3">
<h3 id="adding-environmental-variance">Adding Environmental Variance<a class="anchor" aria-label="anchor" href="#adding-environmental-variance"></a>
</h3>
<p>The genotype-by-environment effects model described above is altered
to model a founder population with a non-zero environmental variance.
All of the above equations remain relevant under the altered model and
only the sampling distributions of the effects in the equations are
changed. Specifically, the variance of environmental covariate is set to
the environmental variance
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>σ</mi><mi>E</mi><mn>2</mn></msubsup><annotation encoding="application/x-tex">\sigma_E^2</annotation></semantics></math>)
and the genotype specific slope is set to a different mean and variance.
The mean of the genotype specific slope is set to one and the variance
is set to the genotype-by-environment interaction variance divided by
the environmental variance
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mfrac displaystyle="false"><msubsup><mi>σ</mi><mrow><mi>G</mi><mi>E</mi></mrow><mn>2</mn></msubsup><msubsup><mi>σ</mi><mi>E</mi><mn>2</mn></msubsup></mfrac><annotation encoding="application/x-tex">\tfrac{\sigma_{GE}^2}{\sigma_E^2}</annotation></semantics></math>).</p>
<p>The logic behind these modifications becomes clear with an
examination of equation (10). The right-hand side of equation (10)
contains three terms. The first term equals zero, because the
environmental covariate has a mean of zero (<em>i.e.</em>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mrow><mo stretchy="true" form="prefix">[</mo><mi>w</mi><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">E[w] = 0</annotation></semantics></math>).
The second term equals the environmental variance, because the mean for
genotype specific slope equals one and the variance of the environmental
covariate equals the environmental variance (<em>i.e.</em>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mrow><mo stretchy="true" form="prefix">[</mo><mi>b</mi><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">E[b] = 1</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>V</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>w</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msubsup><mi>σ</mi><mi>E</mi><mn>2</mn></msubsup></mrow><annotation encoding="application/x-tex">Var(w) = \sigma_E^2</annotation></semantics></math>).
Finally, the value of the last term equals the genotype-by-environment
interaction variance (<em>i.e.</em>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>V</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>w</mi><mo stretchy="true" form="postfix">)</mo></mrow><mi>V</mi><mi>a</mi><mi>r</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>b</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msubsup><mi>σ</mi><mi>E</mi><mn>2</mn></msubsup><mfrac displaystyle="false"><msubsup><mi>σ</mi><mrow><mi>G</mi><mi>E</mi></mrow><mn>2</mn></msubsup><msubsup><mi>σ</mi><mi>E</mi><mn>2</mn></msubsup></mfrac><mo>=</mo><msubsup><mi>σ</mi><mrow><mi>G</mi><mi>E</mi></mrow><mn>2</mn></msubsup></mrow><annotation encoding="application/x-tex">Var(w) Var(b) = \sigma_E^2 \tfrac{\sigma_{GE}^2}{\sigma_E^2} = \sigma_{GE}^2</annotation></semantics></math>).</p>
<p>An examination of equation (10) also shows how both the environmental
and genotype-by-environment interaction variances are populations
specific parameters that are under genetic control. For example, a
population with a higher average genotype specific slope will also have
a higher value for the second term in equation (10). This population
thus have a higher environmental variance even though the environment
itself has not changed. Likewise, the amount of genotype-by-environment
interaction variance depends on the variance for the genotype specific
slope.</p>
</div>
<div class="section level3">
<h3 id="relationship-to-finlay-wilkinson-regression">Relationship to Finlay-Wilkinson Regression<a class="anchor" aria-label="anchor" href="#relationship-to-finlay-wilkinson-regression"></a>
</h3>
<p>AlphaSimR’s model for genotype-by-environment effects is effectively
a biological model for Finlay-Wilkinson regression. Finlay-Wilkinson
regression is a classic technique for analyzing genotype-by-environment
interactions <span class="citation">(Finlay and Wilkinson 1963)</span>.
In its simplest form, an individual’s genotype-by-environment
interaction is reduced to an intercept and a slope. The slope in
Finlay-Wilkinson regression is roughly equivalent to genotype specific
slope in equation (9). The intercept in Finlay-Wilkinson regression is
roughly equivalent to the genetic value of an individual in the target
environment.</p>
<p>Initially, the similarity between these two models was not
intentional. The model used in AlphaSimR originates with the model used
by Gaynor <em>et al.</em> <span class="citation">(2017)</span>. This
model treated each QTL’s additive effect as random variable whose value
depended on the value of an environmental covariate. That model is
roughly equivalent to AlphaSimR’s model when the founder population has
zero environmental variance. The difference between the models is that
AlphaSimR scales effects to exactly achieve a desired variance and the
Gaynor <em>et al.</em> model randomly sampled effects so that the
expectation of the sampled effects equaled the desired variance. It was
only realized later that a slight change to the model, as described
above, could introduce a specified amount of environmental variance and
that those modifications would result in a model with properties
matching those measured by Finlay-Wilkinson regression.</p>
</div>
<div class="section level3">
<h3 id="user-interaction">User Interaction<a class="anchor" aria-label="anchor" href="#user-interaction"></a>
</h3>
<p>AlphaSimR users do not directly observe nor set the value of the
environmental covariate. Instead, they indirectly set its value by
providing a randomly sampled p-value for each environment. AlphaSimR
then uses this p-value to calculate the appropriate value of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>w</mi><annotation encoding="application/x-tex">w</annotation></semantics></math>
for any level of environmental variance. Users should note that p-values
follow a uniform distribution over the range of zero to one, so it is
recommended that users randomly generate p-values using R’s built-in
<code>runif</code> function. A p-value of 0.5 corresponds to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>w</mi><annotation encoding="application/x-tex">w</annotation></semantics></math>
equaling zero and is equivalent to the target environment. If the user
does not supply a p-value, AlphaSimR will sample one at random.</p>
<p>Genetic values and genetic variances reported by AlphaSimR are always
for the target environment. Since the target environment corresponds to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>w</mi><annotation encoding="application/x-tex">w</annotation></semantics></math>
equaling zero, equation (8) is effectively ignored in these
calculations. Only the values for phenotypes and phenotypic variance
ever reflect a contribution from genotype-by-environment
interaction.</p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-FinlayWilkinson" class="csl-entry">
Finlay, K. W., and G. N. Wilkinson. 1963. <span>“The Analysis of
Adaptation in a Plant-Breeding Programme.”</span> <em>Australian Journal
of Agricultural Research</em> 14 (6): 742–54. <a href="https://doi.org/10.1071/ar9630742" class="external-link">https://doi.org/10.1071/ar9630742</a>.
</div>
<div id="ref-Gallais" class="csl-entry">
Gallais, A. 2003. <em>Quantitative Genetics and Breeding Methods in
Autopolyploid Plants</em>. Paris: INRA.
</div>
<div id="ref-gaynor_2017" class="csl-entry">
Gaynor, R. Chris, Gregor Gorjanc, Alison R. Bentley, Eric S. Ober, Phil
Howell, Robert Jackson, Ian J. Mackay, and John M. Hickey. 2017.
<span>“A Two-Part Strategy for Using Genomic Selection to Develop Inbred
Lines.”</span> <em>Crop Science</em> 57 (5): 2372–86. <a href="https://doi.org/10.2135/cropsci2016.09.0742" class="external-link">https://doi.org/10.2135/cropsci2016.09.0742</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Gaynor.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
